### PR TITLE
SERVER-95889 Make FLE2IndexedEqualityEncryptedValueV2 wrap mc_FLE2IndexedEncryptedValueV2_t

### PR DESCRIPTION
**Pull request by marksg07**
Main repo PR: https://github.com/10gen/mongo/pull/29265

- Moves structs needed by fle_crypto to headers
- Adds validation functions for mc_FLE2IndexedEncryptedValueV2_t and mc_FLE2TagAndEncryptedMetadataBlock_t which are needed by fle_crypto

### CODE REVIEW

File: src/mc-fle2-payload-iev-private-v2.h [hunk 1]
```diff
@@ -80,7 +80,39 @@
  *
  */
 
-typedef struct _mc_FLE2IndexedEncryptedValueV2_t mc_FLE2IndexedEncryptedValueV2_t;
+typedef enum {
+    kTypeInitV2,
```
Comment (sgolemon-corp) [inline src/mc-fle2-payload-iev-private-v2.h:84]: I think you're going to find that even these names are too generic given how C ODRs its enums into the global symbol space.  It wasn't a problem when these were scoped to a single TU, but now that they're in a header, we should give them more distinctness.

e.g.
```c
typedef enum {
  kFLE2IEVTypeInitV2,
  kFLE2IEVTypeEqualityV2,
  kFLE2IEVTypeRangeV2,
};
```

File: src/mongocrypt-util.c [hunk 2]
```diff
@@ -124,6 +124,34 @@ const char *mc_bson_type_to_string(bson_type_t t) {
     }
 }
 
+bool mc_is_valid_bson_type(int bson_type) {
```
Comment (sgolemon-corp) [inline src/mongocrypt-util.c:127]: This function seems.... sus.

You seem to just be checking if the value is in the explicitly list of named numeric values.

Thing is that's not very meaningful in terms of if a type is "valid" or not.

`MINKEY`-`MAXKEY` are all "valid" as BSON types, for example.  Those two constants define the endpoints of the "User Defined BSONElements" range.

On the other side, EOD and UNDEFINED don't make sense as valid encodable types. (Though "Undefined" is defined, it's just no longer a valid type, so that one gets.... complicated).

TL;DR - The only truly "invalid" type is EOD.

Unless I'm misunderstanding the purpose of this method, I'd probably just reduce it to:
```c
return bson_type != BSON_TYPE_EOD;
```

**Review by sgolemon-corp — CHANGES_REQUESTED** [review]

File: src/mongocrypt-util.c (hunk 2 above)
Comment (joshbsiegel) [reply src/mongocrypt-util.c:127]: I went ahead and modified this in order to get a test passing, feel free to change it back if necessary.

Comment (marksg07) [reply src/mongocrypt-util.c:127]: Huh, I didn't realize that `[MAXKEY, MINKEY]` are valid, I was wondering what those meant. Thanks for the clarification.

File: src/mc-fle2-payload-iev-private-v2.h (hunk 1 above)
Comment (marksg07) [reply src/mc-fle2-payload-iev-private-v2.h]: Changed

**Review by kevinAlbs — APPROVED** [review]
LGTM with minor comments.

**Review by sgolemon-corp — APPROVED** [review]

**Review by markbenvenuto — APPROVED** [review]
